### ISSUE TITLE
Inquiry Regarding Configuring Cache File Paths in Ultralytics Training

### ISSUE THREAD

**Issue by 102757017** [ultralytics/ultralytics#22429]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

specify the storage path for  train.cache and val.cache within model.train()

### Use case

During model training with Ultralytics, the system automatically generates train.cache and val.cache files in the dataset's labels directory to accelerate data loading.

However, when training in the Kaggle environment, the dataset directory has read-only permissions, which causes the training process to fail. While copying the dataset to the working directory can serve as a temporary workaround, it consumes additional space and becomes cumbersome when updating the dataset.

Therefore, I would like to inquire if it's possible to specify the storage path for cache files (such as train.cache and val.cache) within model.train(), in order to avoid the permission issue?

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @102757017, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving quickly—an Ultralytics engineer will follow up soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Your question about configuring cache file paths for training is noted. To help us assess this request and route it correctly, please share a bit more context:
- Your exact `ultralytics` version and environment details (e.g., Kaggle runtime, GPU/CPU).
- The command or Python snippet you’re using (e.g., `model.train(...)`) and your dataset YAML path.
- A short log excerpt showing the permission error in Kaggle (including the dataset labels directory and the attempted cache path).
- Any constraints or preferences for how you’d like to specify the cache location (e.g., single directory for both train/val, per-split paths, or environment-variable-based).
- Dataset size and typical cache sizes you’re seeing, to understand storage implications.

If this is a 🐛 Bug Report related to cache creation in read-only environments, please also provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question about workflow in Kaggle, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). This will help us determine whether an API enhancement is needed or if there’s an existing configuration that meets your needs.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Custom cache directory isn't supported.

The cached labels are saved so that you don't have to regenerate it every time you train. But it doesn't affect training speed. Each epoch will still have the same speed even if the label cache can't be saved because Ultralytics internally still uses cached labels.

**Comment by Namithnp** [comment]
Cache availability or non-availability does not affect the time it takes for an epoch to run, but the problem is the 12 hour quota on Kaggle. The time Ultralytics takes for scanning a huge dataset is still eating away the compute time. I have tried everything to avoid this without success
1. Symlinks in kaggle/working -> Ultralytics still looks at kaggle/input and again fails to write the .cache file
2. Moving entire dataset to the writable tmp directory -> It still "restores corrupt images" even when .cache is present. And I/O is significantly slow while training using this directory
3. Running 1 epoch locally, writing an ultralytics .cache file and uploading it along with the dataset to Kaggle -> The .cache file gets ignored (possibly due to path or timestamp mismatch). Modifying this .cache file on kaggle is a mess

**Comment by glenn-jocher** [comment]
You're likely hitting cache validation rather than timestamps: `labels.cache` is only reused when both its `version` and dataset `hash` match, and the default `cache_path` still comes from the labels directory, so caches created locally or under a different Kaggle mount are skipped; as a workaround, override `YOLODataset.get_labels()` or `cache_labels()` to redirect `cache_path` to a writable location following the [custom dataset example](https://docs.ultralytics.com/guides/coco-json-training), and please open a new issue for visibility with a minimal repro, the full text traceback, and:
```python
from ultralytics.utils.checks import collect_system_info
collect_system_info()
```

**Comment by Y-T-G** [comment]
> Running 1 epoch locally, writing an ultralytics .cache file and uploading it along with the dataset to Kaggle -> The .cache file gets ignored (possibly due to path or timestamp mismatch). Modifying this .cache file on kaggle is a mess

Is there any warning for it? Where did you place it? It should be inside `labels` folder. Not inside `train` or `val`.
